all-sdk/autotuning.md

```diff
@@ -93,77 +93,70 @@ trait SlotSupplier {
-  async fn reserve_slot(&self);
+  async fn reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>);
 
   /// Tries to immediately reserve a slot, returning true if a slot is available.
-  fn try_reserve_slot(&self) -> bool;
+  fn try_reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>) -> bool;
 
   /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
   /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
   /// then the slot can actually be used to work on the task.
   /// 
   /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
   /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
-  /// 
-  /// `info` may not be provided if the slot was never used
-  /// `error` may be provided if an error was encountered at any point during processing
-  ///     TODO: Error type should maybe also be generic and bound to slot type
-  fn mark_slot_used(&self, info: Option<&Self::SlotKind::Info>, error: Option<&anyhow::Error>);
+  fn mark_slot_used(&self, info: <Self::SlotKind as SlotKind>::Info<'_>);
 
-  /// Frees a slot.
-  fn release_slot(&self, info: &SlotReleaseReason<Self::SlotKind::Info>);
-  
-  /// If this implementation knows how many slots are available at any moment, it should return that here.
-  fn available_slots(&self) -> Option<usize>;
+  /// Frees a slot. This is always called when a slot is no longer needed, even if it was never marked as used.
+  /// EX: If the poller reserves a slot, and then receives an invalid task from the server for whatever reason, this
+  /// method would be called with [SlotReleaseReason::Error].
+  fn release_slot(&self, info: SlotReleaseReason);
 }
 
-struct WorkflowSlotsInfo {
-  used_slots: Vec<WorkflowSlotInfo>,
-  /// Current size of the workflow cache.
-  num_cached_workflows: usize,
-  /// The limit on the size of the cache, if any. This is important for users to know as discussed below in the section
-  /// on workflow cache management.
-  max_cache_size: Option<usize>,
-  // ... Possibly also metric information
-}
-struct ActivitySlotsInfo {
-  used_slots: Vec<ActivitySlotInfo>,
+/// This trait lets implementors obtain other information that might be relevant to their decision on whether to hand
+/// out a slot. It's a trait rather than a struct because the most up-to-date information should be available even
+/// if waiting for some time in the blocking version of `reserve_slot`.
+pub trait SlotReservationContext<SK: SlotKind>: Send + Sync {
+  /// Returns information about currently in-use slots
+  fn used_slots(&self) -> &[SK::Info];
+  
+  // ... anything else users end up wanting here
 }
-struct LocalActivitySlotsInfo {
-  used_slots: Vec<LocalActivitySlotInfo>,
+
+pub enum SlotReleaseReason {
```

Comment:
If the worker is shutting down is that an `Error` or `NeverUsed`?

Reply:
I would say `NeverUsed`, since if it _is_ used, then we're waiting until that task completes before finishing shutdown. (or, if we're not, then it'd be an `Error`... but we should be waiting for tasks to finish)

Reply:
Fine with me, might be worth commenting so we are consistent across SDKs.